src/rust_impl/fs/s3.rs

```diff
@@ -0,0 +1,302 @@
+        }
+        let bucket_name = parts[0];
+        let key = parts[1];
+
+        // println!("in download fn with: bucket_name: {bucket_name} & key: {key}");
+        // Fetch the object from S3
+        let response = self
+            .s3client
+            .get_object()
+            .bucket(bucket_name)
+            .key(key)
+            .send()
+            .await?;
+
+        // Read the response body
+        let mut stream = response.body.into_async_read();
+        let mut file = File::create(local_path).await?;
+        tokio::io::copy(&mut stream, &mut file).await?;
+
+        Ok(())
+    }
+
+    async fn byte_range_download(
+        &self,
+        remote_path: &str,
+        local_path: &str,
+        num_threads: usize,
+    ) -> Result<(), Box<dyn std::error::Error>> {
+        // Extract bucket & key
+        let parts: Vec<&str> = remote_path
+            .trim_start_matches("s3://")
+            .splitn(2, '/')
+            .collect();
+        if parts.len() < 2 {
+            return Err("Invalid S3 path format. Expected 's3://bucket-name/key'".into());
+        }
+        let bucket_name = parts[0];
+        let key = parts[1];
+        // println!("in byte-range download fn with: bucket_name: {bucket_name} & key: {key}");
+        // Step 1: Get file size
+        let head_response = self
+            .s3client
+            .head_object()
+            .bucket(bucket_name)
+            .key(key)
+            .send()
+            .await?;
+        let file_size = head_response.content_length().unwrap_or(0) as u64;
+
+        // Step 2: Create an empty file with correct size
+        let file = OpenOptions::new()
+            .create(true)
+            .write(true)
+            .truncate(true)
+            .open(local_path)
+            .await?;
+        file.set_len(file_size).await?; // Preallocate space
+
+        // Step 3: Spawn async tasks for each byte range
+        let chunk_size = file_size / num_threads as u64;
+        let file = Arc::new(Mutex::new(file)); // Shared mutable file
+        let mut tasks = JoinSet::new();
+
+        for i in 0..num_threads {
+            let start = i as u64 * chunk_size;
+            let end = if i == num_threads - 1 {
+                file_size - 1
+            } else {
+                (start + chunk_size) - 1
+            };
+            let s3client = self.s3client.clone();
+            let bucket_name = bucket_name.to_string();
+            let key = key.to_string();
+            let file = Arc::clone(&file);
+
+            tasks.spawn(async move {
+                let range_header = format!("bytes={}-{}", start, end);
+                let response = s3client
+                    .get_object()
+                    .bucket(bucket_name)
+                    .key(key)
+                    .range(range_header)
+                    .send()
+                    .await?;
+
+                let body = response.body.collect().await?;
```

Comment:
Here, we could just keep that in memory for now and have another method to let python access it.